src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringConcat.cs

```diff
@@ -203,6 +307,9 @@ private bool TryExtractStringConcatArgs(BoundExpression lowered, out ImmutableAr
                     }
 
                     break;
+
+                case BoundSequence { SideEffects.Length: 0, Value: BoundCall sequenceCall }:
+                    return TryExtractStringConcatArgs(sequenceCall, out arguments);
```

Comment:
>return TryExtractStringConcatArgs(sequenceCall, out arguments); [](http://example.com/codeflow?start=20&length=63)

How do we know that all locals in this sequence can be dropped?

Reply:
Isn't the fact that `TryExtractStringConcatArgs` will only succeed for a well-known shape of a call and its arguments enough to state that? If no, how we can additionally ensure that and how that can be tested?

Reply:
> Isn't the fact that ```TryExtractStringConcatArgs``` will only succeed for a well-known shape of a call and its arguments enough to state that?

Aren't locals, used by the sequence, a part of the well-known shape? Is sequence a node that nobody else creates and will not ever create it in the future? Why do we assume this is "our" sequence?

> If no, how we can additionally ensure that 

Here is a quote from the [https://github.com/dotnet/roslyn/pull/70971#issuecomment-1886329321](https://github.com/dotnet/roslyn/pull/70971#issuecomment-1886329321):

> a sequence without side effects, **locals are chars that are used in arguments** that are new ReadOnlySpan<char>(in char) calls, other arguments are _String__op_Implicit_ToReadOnlySpanOfChar calls


> and how that can be tested?

At the moment, I cannot think of **an easy way** to observe the failure. However, the fact that we cannot easily come up with a "breaking" scenario is not a reason to not verify assumptions.